db-engine/src/main/java/com/splicemachine/db/iapi/sql/dictionary/fk/Dfs.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2012 - 2020 Splice Machine, Inc.
+ * This file is part of Splice Machine.
+ * Splice Machine is free software: you can redistribute it and/or modify it under the terms of the
+ * GNU Affero General Public License as published by the Free Software Foundation, either
+ * version 3, or (at your option) any later version.
+ * Splice Machine is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ * You should have received a copy of the GNU Affero General Public License along with Splice Machine.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.splicemachine.db.iapi.sql.dictionary.fk;
+
+import com.splicemachine.db.iapi.error.StandardException;
+import com.splicemachine.utils.Pair;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Stack;
+
+public class Dfs {
+    private final Graph graph;
+    private final String newConstraintName;
+    int[] dfsParent;
+    boolean[] processed;
+    boolean[] discovered;
+    private boolean trackParents;
+
+    enum EdgeClassification {TREE, BACK, FORWARD, CROSS};
+    EdgeClassification[] edgeTypes;
+    Stack<Integer> stack;
+    int[] entryTime;
+    int[] exitTime;
+    int time;
+    ArrayList[] parents;
+    boolean finished;
+
+
+    public Dfs(Graph g, String newConstraintName) {
+        this.graph = g;
+        dfsParent = new int[g.getVertexCount()];
+        processed = new boolean[g.getVertexCount()];
+        discovered = new boolean[g.getVertexCount()];
+        edgeTypes = new EdgeClassification[g.getVertexCount()];
+        entryTime = new int[g.getVertexCount()];
+        exitTime = new int[g.getVertexCount()];
+        init();
+        stack = new Stack<Integer>();
```

Comment:
A Deque is the preferred Java Stack type (Stack is based on Vector, which is synchronized by default and the API is not as consistent as Deque)

Reply:
done.